Sources/NIO/System.swift

```diff
@@ -310,10 +310,22 @@ internal enum Posix {
             let fd = sysAccept(descriptor, addr, len)
 
             #if !os(Linux)
-                if fd != -1 {
-                    let ret = try? Posix.fcntl(descriptor: fd, command: F_SETNOSIGPIPE, value: 1)
-                    assert(ret == .some(0),
-                           "unexpectedly, fcntl(\(fd), F_SETFL, F_SETNOSIGPIPE) returned \(ret.debugDescription)")
+                do {
+                    // Result is uninteresting, fcntl() only returns nonzero on error.
+                    if fd != -1 {
+                        _ = try Posix.fcntl(descriptor: fd, command: F_SETNOSIGPIPE, value: 1)
+                    }
+                } catch let err as IOError {
+                    // Per setsockopt(SO_NOSIGPIPE), fcntl(F_SETNOSIGPIPE) can return EINVAL if
+                    // the remote closed the connection after accept() returned. This is not
+                    // programmer error and thus should not be covered by an assertion.
+                    assert(err.errnoCode == EINVAL,
+                           "unexpectedly, fcntl(\(fd), F_SETFL, F_SETNOSIGPIPE) returned \(err)")
+                    
+                    // Throw on EINVAL but silently ignore other errors when assertions are disabled.
+                    if err.errnoCode == EINVAL {
+                        throw err
```

Comment:
After talking a bit with @Lukasa we both agreed that we should just always throw and not "discard" errors in some cases. I really dont remember why we even ignored these at the first place.

Reply:
Done in latest commits.